typing/ctype.ml

```diff
@@ -266,77 +266,84 @@ type equations_generation =
   | Forbidden
   | Allowed of { equated_types : TypePairs.t }
 
-type unification_mode =
-  | Expression (* unification in expression *)
+type unification_environment =
+  | Expression of
+      { env : Env.t;
+        in_subst : bool; }
+    (* unification in expression *)
   | Pattern of
-      { equations_generation : equations_generation;
+      { env : Env.t ref;
+        equations_generation : equations_generation;
         assume_injective : bool;
-        allow_recursive_equations : bool; }
+        allow_recursive_equations : bool;
+        gadt_equations_level : int;
+        unify_eq_set : TypePairs.t; }
     (* unification in pattern which may add local constraints *)
-  | Subst
-    (* unification during type constructor expansion; more
-       relaxed than [Expression] in some cases. *)
 
-let umode = ref Expression
+let get_env = function
+  | Expression {env} -> env
+  | Pattern {env} -> !env
 
-let in_pattern_mode () =
-  match !umode with
-  | Expression | Subst -> false
+let set_env uenv env' =
+  match uenv with
+  | Expression _ -> invalid_arg "Ctype.set_env"
+  | Pattern {env} -> env := env'
+
+let in_pattern_mode = function
+  | Expression _ -> false
   | Pattern _ -> true
 
-let in_subst_mode () =
-  match !umode with
-  | Expression | Pattern _ -> false
-  | Subst -> true
+let order_type_pair t1 t2 =
+  if get_id t1 <= get_id t2 then (t1, t2) else (t2, t1)
 
-let can_generate_equations () =
-  match !umode with
-  | Expression | Subst | Pattern { equations_generation = Forbidden } -> false
+let add_type_equality uenv t1 t2 =
+  match uenv with
+  | Expression _ -> invalid_arg "Ctype.add_type_equality"
+  | Pattern r -> TypePairs.add r.unify_eq_set (order_type_pair t1 t2)
+
+let unify_eq uenv t1 t2 =
+  eq_type t1 t2 ||
+  match uenv with
+  | Expression _ -> false
+  | Pattern r -> TypePairs.mem r.unify_eq_set (order_type_pair t1 t2)
+
+(* unification during type constructor expansion; more
+   relaxed than [Expression] in some cases. *)
```

Comment:
Though I see that this comment is just moved, could you maybe make "in some cases" more explicit? On the call this morning, you had a nice description of what this means, so putting that here -- or on the `in_subst` field declaration -- would be helpful.